### PR TITLE
feat: complete queries, make overall project look pretty #1

### PR DESCRIPTION
**Pull request by amaanq**
Same as last time haha, I was a bit lazier this time as fixing/debugging the grammar took a bit longer. I followed convention with regards to modifying the structure of precedences, iirc tree-sitter-python and c or c++ look similar to that.

Let me know of anything to edit/fix! :) 

Also, same plan as last time (I hope?), I'd like to be added as a contributor again... Yuck was on my TODO list but other languages and school have distracted me. Thanks for taking the time to wriite this!

### CODE REVIEW

**Comment by Philipp-M** [comment]
Thanks for all the work you've put into this!

Well it would be a little bit easier to review, when this would've been split up across different PRs (especially the changes in the parser)...
Also to avoid work for both of us.

Anyway...

I think the parser before this PR was pretty much complete, maybe needed some code beautification, but should completely describe the AST of any yuck file, well if no bugs were in there, but I tested parsing on quite a few yuck files and it produced an AST that is close to the AST of the main repo.

In general I would prefer to stay close to the parsers in the main eww repo (e.g. yuck has an embedded DSL called "simplexpr") , mainly:

* https://github.com/elkowar/eww/blob/master/crates/yuck/src/parser/parser.lalrpop
* https://github.com/elkowar/eww/tree/master/crates/yuck/src/parser
* https://github.com/elkowar/eww/blob/master/crates/simplexpr/src/simplexpr_parser.lalrpop
* https://github.com/elkowar/eww/tree/master/crates/simplexpr/src/parser

This is to easily adjust to changes in the syntax and to maintain an AST that is close to the one used in the main repo.
That is why e.g. `symbol` or `simplexpr` is used as node names. E.g. in `(for entry in my-json ...)` `for` is not really a widget, symbol makes probably more sense here (which is used in the main repo as well).

To be able to highlight specific nodes differently (i.e. the `include` node you have specified)  I think it may be better to match these via e.g. `(#match? @constant "^include$")` etc. and maintain an AST that describes the language well (well it's a syntax tree after all)

I've seen that you changed quite a few regexes (e.g. identifier), as previously stated, I think it's better to keep the current regexes, because they are mostly copied from the original parser/lexer (e.g. for the symbol https://github.com/elkowar/eww/blob/6a64a286291456984011c4fb160def106ee55dff/crates/yuck/src/parser/lexer.rs#L68), so they closely represent what the language is really be able to do.

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: Why change this to something different than `/./`? If looking at the original parser here: https://github.com/elkowar/eww/blob/6a64a286291456984011c4fb160def106ee55dff/crates/simplexpr/src/parser/lexer.rs#L211

The original parser ignores any char, so this is actually not correct. In fact it just skips any escaped chars at all (doesn't introduce anything special in the AST), but to be able to highlight this, I've introduced a new node `escape_sequence` to be able to specify highlighting for escape sequences. If any of the specified characters should be highlighted differently, this should likely be done with query highlights.

I have also originally thought about copying the escape_sequence from something like javascript, but the parser of `simplexpr` doesn't handle textual encoded unicode characters such as `/u[0-9a-fA-F]{4}/`, it just skips one character (the next after the escaping)

Comment (amaanq) [reply grammar.js]: I see, I had never taken a look at the original parser, sorry about that!! 

So, if I'm understanding you right, only one-char escape sequences should be highlighted? Like \n, \b, \t, etc..

**Comment by amaanq** [comment]
I've updated parts of the grammar to remove the include rule and use regex matches to highlight def keywords and include keywords, as well as restoring symbol and its original regex. Anything more to restore? I agree with keeping it in-line as much as possible with the DSL, but tree-sitter grammars usually aren't a 1-to-1 mapping (and for some languages that's not possible :))

**Comment by amaanq** [comment]
The only issue now is in this screenshot:

image

"used_mem_perc" cannot be highlighted as a field/property, because all of `EWW_RAM.used_mem_perc` is matched as a symbol. I'm not sure if we can use a regex to highlight it, but that'd be nice imo :) unless you think it's not *meant* to be a field.

**Comment by Philipp-M** [comment]
Well, the parser is a little bit better, but overall (what I wanted to say in my previous comment) is that the parser was already done, and represents the AST already well. So IMHO the only thing that might be subject to change in the parser/grammar (apart of course of new language features added in the main repo sometime in the future) is probably a little cleanup (like prec naming, and maybe adjusting prec (I for whatever reason started with 10 and counted down to 4, so maybe subtract 3, so that the highest precision is 7).
But the overall logic of the parser (including most names of the nodes) should IMHO stay. Right now, I think only the precisions in the `simplexpr` should probably be named and grouped together (something like `PREC.json_access`)

Sorry to say that, but right now I see a lot of issues in your parser, e.g. there's no float in the original parser (only one number node), there are a few operators not supported in yuck (e.g. bitwise operators), "simplexpr" is not correctly represented wherever it's used (i.e. in string interpolation, or as you noted, no "simplexpr" in your block node, see below) etc.

> "used_mem_perc" cannot be highlighted as a field/property, because all of EWW_RAM.used_mem_perc is matched as a symbol.

Can you please try the original parser? This should've already worked there (it's an `index` node there, though the name could probably be changed to something different...)

I really recommend to read the lalrpop parser which I've linked. Looking at the real parser is IMHO generally a good way to start a tree-sitter parser. You mostly just have to "translate" the code, you don't have to guess that much, and you're learning a lot while doing it, e.g. the lalrpop parser, or (simple) recursive descent parsing in the case of ungrammar etc. I think it also will likely represent the real grammar of the language more closely. 

The grammar should really just describe the AST of the real language, no special handling of nodes etc. (if it's not in the real language), only if it really makes sense (e.g. a normal regex in the queries is not enough, but I think this happens very rarely). One specific example in the case of the yuck parser I wrote is, that the original parser parses also negative numbers (and `+` signs): Compare the original regex(of the real parser): `[+-]?(?:[0-9]+[.])?[0-9]+` with `(?:[0-9]+[.])?[0-9]+` of my parser, I have instead included `+` as unary operator to be able to highlight it correctly, and be able to represent it in the (tree-sitter) AST better.

I *think* the most effective changes should be done in the highlights queries instead (and generally other queries, I have not really invested much time there as you probably noticed).
E.g. if you want to differ syntax highlighting between floats and integers (whether that makes sense or not), you should probably check, if the number has a dot in it (with a `match` and regex probably), or if you want to highlight some symbols differently (e.g. all the builtins, like `defwidget` or `box` or `for` etc.) like this: https://github.com/cstrahan/tree-sitter-nix/blob/1b69cf1fa92366eefbe6863c184e5d2ece5f187d/queries/highlights.scm#L21

Sorry if that came a bit harsh, that's not my intention, apart from the parser most is fine at first look, and I appreciate all the other changes.

**Comment by amaanq** [comment]
Sorry I got caught up...I more or less kept it as the original parser. Some notes:

I don't really like calling ast `$._ast`, because it defines a scope and should be captured as so. I think `$.ast` also gives no hint or clue to a user looking at a tree what it actually is, so naming it `$.ast_block` is better imo

I am **very** against clumping floats and integers together, even if it is in the DSL. This might seem like a minor nitpick but using regex-based queries to distinguish them is unnecessarily costly. (I like your idea of putting +- under unary ops to highlight them properly)

I renamed property to keyword and primary_expression to simplexpr, but I think the expression block I added should stay, it adds clarity and putting it as a supertype ensures it doesn't show up in the CST. 

A thing to note is, tree-sitter generates CSTs not ASTs. In this context it's trivial since so does the lalrpop parsers practically do too. But it's a distinction to keep in mind going forward :)

I do like adhering to the original parser *as much as reasonably possible*, but exceptions are ok when warranted, and naming doesn't really have to be *exactly* same (think ast -> ast-block).

Everything else seems ok to me now, thanks for bearing with me! I never even looked at the lalrpop files when I first made these changes, so thanks for linking those

**Comment by amaanq** [comment]
Wow, really sorry about the binops! Completely forgot to remove them...I copied over the binary_expression from python or c, it's always my go-to so I'm sorry about that, removed now

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: I'm not against introducing a distinction between floats and integers per se, but as said previously, a CST should be a superset of the AST, thus, I would group in that case an `integer` and a `float` node together to a `number` node (`number: => choice($.integer, $.float),`)

File: grammar.js
Comment (Philipp-M) [reply grammar.js]: yes, please just use my `escape_sequence`, it's more precise and shorter (one liner).

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: Why making unnamed nodes here? This should be represented in the syntax tree. We can alias it, so that there's no distinction between the different fragments (backtick, double quote etc.), but it *should* be visible in the syntax tree. (Same applies for the other strings below). Also why remove the string function? This IMHO unnecessarily clutters the code.

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: This is not really an expression AFAIK (once again, a CST *should* be a superset of the AST).

The *only* real expression there is in the yuck language is the DSL `simplexpr` which was already described by my `expr: $ => seq("{", $.simplexpr, "}")`, and while the name of the node still might not be completely true, (as it's a braced block that *contains* an expression) we could probably rename that node to something like `expression_block` or something alike. But the rest of the nodes you put there should stay IMHO their own top-level rule as described in the original AST.

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: I'm fine with giving the `_ast` rule a concrete name like `ast_block` but it should contain the same as it originally did. i.e. like `ast_block: $ => choice($.list, $.array, $.keyword, $.symbol, $.literal, $.string, $.expression_block),`

Otherwise it's wrong, and not really representing the real AST. The real AST allows top level arrays and something like that (check the lalrpop parser).

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: Why alias this? IMHO that makes queries unnecessarily more complex (you have to differ between interpolation_expression and simplexpr although it's the same), just use simplexpr, you can match with a query whether it's in a interpolation context or in a `expression_block`

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: Please cover this *exact* regex with the `integer` and `float` node, otherwise there are some expressions that are legal in yuck, but not in the tree-sitter grammar (like `1.`) or are legal in this parser, but aren't in yuck, like `.3`. *This* is why I took the original syntax as is, because I think it's overkill to differ between integers and floats in the tree-sitter AST as it's not handled differently in the language anyway (though as below described, you can do that but it should cover this *exact* regex).

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: I don't think `supertypes` is needed at all. The specified (super-)rules are visible in the tree anyway (and should IMHO stay like that, so that it's obvious that the DSL `simplexpr` is used).

File: grammar.js
Comment (Philipp-M) [inline grammar.js:163]: see above regarding strings

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: Just a tip, I think it's easier to read an extra function that's called `binary_expr` (as before) inside the `choice`, instead of some magical functional programming in js (don't get me wrong, I really like functional programming, but I think it should be obvious at first sight if possible what is happening here if it's possible.

Anyway I'm still fine with that, as it's just marginally more complex...

File: grammar.js
Comment (Philipp-M) [inline grammar.js:46]: I don't *think* that's necessary, but on the other hand it probably doesn't harm to keep it...

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: This shouldn't be a symbol, just the ident it was before, it's not a symbol...

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: As said previously this should probably be renamed to `expression_block` (it's a `simplexpr` in the original parser, but that name is already taken... it could also be called `simplexpr_block` as it's probably even more precise)

File: grammar.js
Comment (Philipp-M) [reply grammar.js]: What do you mean with "optional preceding :"? 
`(?:)` is a non-capturing group, there's no optional : in there

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: string is not a literal...

literals should really be like constants (things that don't need to be evaluated at all), but there is no string literal in this language (at least it's not specially handled). If you want to highlight "string literals" differently you should do it in a query instead (i.e. a string containing no string interpolation).

File: grammar.js
Comment (Philipp-M) [reply grammar.js]: also idents have a different regex, please refer to the lalrpop parser

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: This should have the highest precision instead of the lowest, though I don't think it's necessary at all, as it's kind of an atom rule anyway and it will be parsed correctly anyway.

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: This and for all the other json access access rules, the precision is missing, I think without the precision the AST is wrong, as this is kind of a binary expression, and `*` will have a higher precision than `.` or `?.`.

Thinking about it, I think also the "json atoms" (`json_array` and `json_object`) need to keep the precision for unambiguous correct parsing, and for the sake of completeness, I would also keep the parenthesized_expression precision (although I think it would parse unambiguously without it as well).

Again, keeping close to the original parser will avoid headaches and bugs introduced unnecessarily, only if you're absolutely sure that something is unnecessary, you may remove it.

File: grammar.js
Comment (Philipp-M) [inline grammar.js]: See in a different comment, these should be directly at the top-level `ast_block`

File: grammar.js
Comment (Philipp-M) [reply grammar.js:46]: But I'm not sure, as there's also an ident rule in the simplexpr

File: grammar.js
Comment (Philipp-M) [inline grammar.js:18]: I'm not sure if it's a good idea to use 0 (as this is AFAIK the default precision for everything else), I would use 1 instead, though it may be indifferent probably...

File: grammar.js
Comment (Philipp-M) [reply grammar.js]: Also I would make either `_literal` a named node, or keep `boolean_literal` or `number_literal` as I think it's more descriptive to have literal somewhere, if it's indeed a literal.

**Review by Philipp-M — CHANGES_REQUESTED** [review]
> A thing to note is, tree-sitter generates CSTs not ASTs. In this context it's trivial since so does the lalrpop parsers practically do too. But it's a distinction to keep in mind going forward :)

Yes indeed, but CSTs are usually a superset of an AST (i.e. containing *all* the nodes of the AST and more, in the best case being able to losslessly reconstruct the original source file, as can e.g. be done via [rowan](https://github.com/rust-analyzer/rowan) (used in quite a few parsers in rust like rust-analyzer, rome or rnix )).

In your parser that's not exactly the case, as a few nodes are called differently (which I'd like to avoid), and are sometimes simplifying nodes where it's not appropriate.

Well as said before, there are quite a few issues in the parser (the ones I listed before were just examples), and it would've been less work for both of us, if the current/previous parser would've just been "refined" instead of almost rewritten (in its core logic)...
Anyway I reviewed the parser (but only that for now) and added quite a few comments, I *hope* I covered everything, but I think it's necessary to carefully test this with all kinds of edge cases. I hope you understand now, why I recommend to *closely* follow the real parser (to depict the *same* syntax tree as the language really supports).

Honestly I think it's easier to just "restart" based on my parser and refine it instead of rewriting so much of the logic (which as noted in the comments, introduced quite a few bugs, which weren't there before).

> (and for some languages that's not possible :))

By the way as I overlooked that previously, AFAIK tree-sitter is indeed general enough (GLR algoritm) to parse every language, you sometimes have to use an escape hatch (like the scanner used in here), and sometimes (though rarely) parsers will get ugly (a handwritten recursive descent parser may be easier to read and write), but it *should* be possible to cover them.

File: grammar.js
Comment (amaanq) [reply grammar.js:18]: More about being relative

Also removed parenthesized, that was unnecessary

File: grammar.js
Comment (amaanq) [reply grammar.js]: Parenthesized is unnecessary, but I'm confused as to how this will have conflicts/issues as it doesn't need that level of precendence comparatively. Keeping it close to the original parser is fine, but there's no conflicts without it

File: grammar.js
Comment (amaanq) [reply grammar.js]: Yep, fixed

File: grammar.js
Comment (amaanq) [reply grammar.js]: renamed it to expr, that's fine with me

File: grammar.js
Comment (amaanq) [reply grammar.js]: aliased the string fragments to all be `$.string_fragment`

I found the string function to be not very readable, and sacrificed readability for brevity. I find this better as we can see there's three different types of strings in the `$.string` rule

File: grammar.js
Comment (amaanq) [reply grammar.js]: Yeah I moved integer and float to number

**Comment by amaanq** [comment]
I've sort-of redone it based on the original parser now, sorry again for misunderstanding your intentions at first.

I like how I handle escape sequences, because "escapes" such as '\A' are not really an escape sequence, and thus won't be processed as one, but '\n', '\a', etc will be. This is an important distinction in editors and, more or less, in general

**Comment by Philipp-M** [comment]
> I like how I handle escape sequences, because "escapes" such as '\A' are not really an escape sequence, and thus won't be processed as one, but '\n', '\a', etc will be. This is an important distinction in editors and, more or less, in general

Well I agree that they might be highlighted differently (like you said only escape sequences that make sense, the rest just as string literal), but IMHO they should be represented correctly in the tree still (I think your parser fails with valid escape sequences of the language, which would be `\A` as example. As said, I'd keep the escape_sequence so that valid yuck can be parsed correctly, but add different highlighting for the right ones (e.g. via a match query).

Can you please check each of my comments (they took some time to write/and review)?

At a quick look a lot of my comments were not yet considered yet in the new version...

File: grammar.js
Comment (Philipp-M) [reply grammar.js]: I haven't tested it, but think about something like:

```
foo+bar.baz.abc
```

I think the precedence is something like this without explicit precedence:

```
(foo+bar).baz.abc
```
 
while it should be, according to the real grammar:
```
foo+(bar.baz.abc)
```

You can maybe write a test for that.

I think something similar happens, if you combine an array with either `json_access` or `json_safe_access`

**Comment by amaanq** [comment]
Sorry about that, I didn't notice the "8 hidden conversations" button showing a lot more of your comments, let me get on that now..

As for escape sequences, I don't think \A should even be parsed as an escape sequence to begin with...as it's really *not* an escape sequence. AFAIK some grammars distinguish them, some don't, so it's more about "stick to the DSL" or "make editor highlighting easier". Using regex is costly and it'd just be easier to throw all escape_sequences under `@string.escape`. I'll let you make the final call on that, just thought that I should say my two cents about it.

File: grammar.js
Comment (amaanq) [reply grammar.js]: Fixed

File: grammar.js
Comment (amaanq) [reply grammar.js]: I'll remove simplexpr, but ast_block and literal are ok there imo

File: grammar.js
Comment (amaanq) [reply grammar.js]: I think this maintains more clarity having a simple table with the operators and precedence, vs calling a function n times. It might seem unclear at first for someone new to tree-sitter/js, but this is a very standard way of doing things

File: grammar.js
Comment (amaanq) [reply grammar.js:46]: yeah, that is a bit of an issue, but it's good for extensions to know where the beginning/end of a "word" is, typically this is done with identifiers but symbol makes more sense (and ident causes a bug)

File: grammar.js
Comment (amaanq) [reply grammar.js]: oh interesting, thanks I'll add that back

**Comment by amaanq** [comment]
Ok, addressed everything but escape_sequences now, that's your final call

File: grammar.js
Comment (amaanq) [reply grammar.js]: I see, I updated float to make digits after a . optional

File: grammar.js
Comment (Philipp-M) [reply grammar.js]: please also keep the json_array and json_object precedence, just to be on the safe side (see lalrpop parser, one higher than json_access, it's on the same level as the atoms, I think there may still might be cases that are parsed incorrectly)

Comment (amaanq) [reply grammar.js]: done

File: grammar.js
Comment (Philipp-M) [reply grammar.js:46]: It does cause a bug? Is the bug gone, if this is removed?

File: grammar.js
Comment (Philipp-M) [reply grammar.js:163]: I've implemented it slightly differently now

File: grammar.js
Comment (Philipp-M) [reply grammar.js]: Well, I'm fine with that, but does it bring `any` benefit?

**Comment by Philipp-M** [comment]
> so it's more about "stick to the DSL" or "make editor highlighting easier"

Well it turns out it's neither, I've checked the original parser in more detail again, and it seems escaped nodes are just ignored and string literal fragments are combined together. I've implemented that behavior now, with your logic for escape_sequences. I have also fixed the number node, which comment you seem to have overlooked (the behavior was a little bit different anyway (yuck only understands either `12.0` as float or `12` as integer, nothing else (missing leading zeros or no decimal places etc.))).

I have also provided a (eslint) config for prettier and reformatted the project (as prettier is kind of a standard). For the future, a tip: please don't reformat code at the same time when features are included, use a reformat PR or something like that. And in general only reformat to a different style, if the maintainers agree, reformatting can possibly create a lot of merge conflicts with different PRs (not the case here but anyway...).

The parser is now ok IMHO.
I have not yet tested/reviewed the highlight queries yet (or the other things added here)

File: grammar.js
Comment (amaanq) [reply grammar.js:46]: It causes a bug if `$.ident` was used for word instead of `$.symbol`, not exactly sure but I think it's because tree-sitter will prefer to match the word rule over other regexes..

**Comment by amaanq** [comment]
> a tip: please don't reformat code at the same time when features are included, use a reformat PR or something like that

Yes, this is a habit I need to break but having your editor automatically do it makes it a bit of a nuisance...I'll work on that :)

Let me test your changes real quick..

Big fan of that string refactor, thanks!

**Comment by Philipp-M** [comment]
> Big fan of that string refactor, thanks!

Ok good to hear :)

I've also slightly reordered the rules, so that they read more linearly

**Comment by amaanq** [comment]
FYI, that reordering of node rules is unconventional, lexical precedence (which should be documented better imo), takes place based on the order of the rules, which is why in almost every grammar you have comments at the bottom and idents & literals near the bottom as well. I think for **this** case it's ok, but it's something to be aware of

Tests are happy so this LGTM now as well (well by test I mean the one example in the eww repo)

File: grammar.js
Comment (amaanq) [reply grammar.js]: Capturing these nodes in queries while not having them show up in the tree-sitter tree output is the benefit iiirc

From the docs:

Supertype Nodes
In Tree-sitter grammars, there are usually certain rules that represent abstract categories of syntax nodes (e.g. “expression”, “type”, “declaration”). In the grammar.js file, these are often written as [hidden rules](http://tree-sitter.github.io/tree-sitter/creating-parsers#hiding-rules) whose definition is a simple [choice](http://tree-sitter.github.io/tree-sitter/creating-parsers#the-grammar-dsl) where each member is just a single symbol.

Normally, hidden rules are not mentioned in the node types file, since they don’t appear in the syntax tree. But if you add a hidden rule to the grammar’s [supertypes list](http://tree-sitter.github.io/tree-sitter/creating-parsers#the-grammar-dsl), then it will show up in the node types file, with the following special entry:

  "subtypes" - An array of objects that specify the types of nodes that this ‘supertype’ node can wrap.

Supertype nodes will also appear elsewhere in the node types file, as children of other node types, in a way that corresponds with how the supertype rule was used in the grammar. This can make the node types much shorter and easier to read, because a single supertype will take the place of multiple subtypes.

**Comment by amaanq** [comment]
Example screenshot with highlights..
image

**Comment by Philipp-M** [comment]
Looks good so far.

I thing, I just noticed on the highlights on your screenshot (and also a thing I've wondered already when looking at the queries) is, that strings are completely highlighted as `@string`. I think it should only be all the quotes/backticks and `string_lit_fragment` shouldn't it?

(see the string interpolation `${music}` )

**Comment by amaanq** [comment]
Oh yes, I think that's an issue with my theme that I never bothered to look into much

See the precedence captures inspected at the music block:

image

(In the nvim-treesitter PR `@embedded` should be removed)

**Comment by amaanq** [comment]
Added a fuzz (for the scanner) and lint action

**Comment by amaanq** [comment]
One nitpick, can we rename string_lit_fragment to string_fragment?

**Comment by amaanq** [comment]
This is good to me now (besides the nitpick)

**Comment by Philipp-M** [comment]
> One nitpick, can we rename string_lit_fragment to string_fragment?

Yeah I think that's fine as well, I've renamed it.

> I think it should only be all the quotes/backticks and `string_lit_fragment`

I've also implemented that, as I think explicit matching is better in this case.

I did a few slight adjustments in the highlight queries, let me know if that is fine for you.
Neovim seems to be quite different in giving labels compared to helix unfortunately, I'll stay with neovim style here I guess.

I changed `symbol` to the label `@tag` as I think it's more appropriate than `@type`

**Comment by Philipp-M** [comment]
Otherwise, I think this is ready to merge (if you're fine with the changes I've committed).

I'll just trust you with the other queries for now, I haven't done much tree-sitter querying apart from syntax-highlighting.

I can add you as collaborator, but please, if you change something in the parser, either wait for my review (I guess that's a good idea anyway), or double-check if the lalrpop parser parses it exactly like the tree-sitter parser.

**Review by Philipp-M — APPROVED** [review]

**Comment by Philipp-M** [comment]
Ok one last small thing I've found in the hightlight queries (matching priority, and a useless/wrong(?) rule for idents)

**Comment by amaanq** [comment]
> I can add you as collaborator, but please, if you change something in the parser, either wait for my review (I guess that's a good idea anyway), or double-check if the lalrpop parser parses it exactly like the tree-sitter parser.

Yeah of course. Any changes to the grammar (breaking or not) I'd submit a PR and consult with you first. The only time I'd merge without consulting is simple stuff like typos, updating a ci/readme, etc.

**Comment by amaanq** [comment]
> Neovim seems to be quite different in giving labels compared to helix unfortunately, I'll stay with neovim style here I guess.

This is something we want to collaborate with helix (and maybe emacs) on because it'd benefit everyone. Just need to get the right push...

**Comment by amaanq** [comment]
> I changed symbol to the label @tag as I think it's more appropriate than @type

Completely forgot about that since practically only html uses it, I agree as well

**Comment by amaanq** [comment]
Check latest commit, I want tags first so any subsequent queries are applied after. 

(It might be that helix does this the opposite way...)

image

Without that commit, any def* and include keywords would be highlighted purple too

But thank you for the edit! Now string interpolations are parsed a bit better for me :) (well I think the issue is the `@variable` for this theme is just pure white or not highlighted at all, thus leading to string (green) being applied anyways in the prior screenshot)

**Comment by Philipp-M** [comment]
> It might be that helix does this the opposite way...

Kind of I guess... and it's probably one more reason why this: 

> This is something we want to collaborate with helix (and maybe emacs) on because it'd benefit everyone. Just need to get the right push...

is even more important. I think helix stops at the first match (haven't looked yet at that part in helix source though, so not sure, but it *feels* like it does, also likely more efficient than neovim btw.), but that introduces even more incompatibilities unfortunately.

I've added one last thing (explicit keyword matching, I've quickly scanned all the keywords that are available in the eww source), I'm gonna merge that now...

**Comment by amaanq** [comment]
I'll likely start a discussion in Helix soon and see what we can get done, I **really** want to get that collaboration/query structure unification solidified...

> I've added one last thing (explicit keyword matching, I've quickly scanned all the keywords that are available in the eww source), I'm gonna merge that now...

If that's the case, `#any-of?` would've been better, sorry for not looking for all matches before

**Comment by Philipp-M** [comment]
> If that's the case, #any-of? would've been better, sorry for not looking for all matches before

You should be able to commit on main now, maybe change that?

**Comment by amaanq** [comment]
Can I just amend to your last commit?

**Comment by Philipp-M** [comment]
sure

**Comment by amaanq** [comment]
One last note for ungrammar and yuck, can you give me `owner` access briefly so I can add the publish.yml action? I'd just add my cargo and npm tokens in the repo secrets so that any tagged commits from 0.0.1 onwards will be auto published to both in case the grammars for either change.

**Comment by Philipp-M** [comment]
> One last note for ungrammar and yuck, can you give me `owner` access

Well that's the issue (as I tried to explain in the issue in ungrammar), I would do that but it's not possible to do without an organization AFAIK.

**Comment by Philipp-M** [comment]
See here: https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-personal-account-settings/permission-levels-for-a-personal-account-repository

**Comment by amaanq** [comment]
> Well that's the issue (as I tried to explain in the issue in ungrammar), I would do that but it's not possible to do without an organization AFAIK.

Aha..I see. Do you have a platform I can dm you on then? Sorry for missing your response in the ungrammar repo

Email works too

**Comment by Philipp-M** [comment]
Sure my email should be in each of the commits ;)

**Comment by amaanq** [comment]
Just emailed

**Comment by Philipp-M** [comment]
> Just emailed

and already set up the secrets, your workflows should work now with the given secrets

**Comment by Philipp-M** [comment]
Btw. do you think I can update the nvim-treesitter PR with the current state or is there still something you want to commit?

**Comment by amaanq** [comment]
Nope it's good now, go for it! Just remove `@embedded`

**Pull request merged**
